### PR TITLE
feat: added a new algorithm to find whether two line segment intersec…

### PR DESCRIPTION
**Pull request by meSajied**
…t or not

#### Checklist

- [ ] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] Sort by alphabetical order
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: geometry/line_segment_intersection.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#include <bits/stdc++.h>
```
Comment (ayaankhan98) [inline geometry/line_segment_intersection.cpp:1]: don't use `<bits/stdc++.h>`

File: geometry/line_segment_intersection.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+using namespace std;
+
+class Point{
```
Comment (ayaankhan98) [inline geometry/line_segment_intersection.cpp:4]: you can use a `struct` if all the members are `public`

File: geometry/line_segment_intersection.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
+#include <bits/stdc++.h>
+using namespace std;
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:2]: fails `cpplint`. must be removed.

File: geometry/line_segment_intersection.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+class Point{
+public:
+    int x, y;
+};
+
+class SegmentIntersection{
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:9]: document your code.

**Comment by kvedala** [comment]
Please see the file [`search/binary_search.h`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/16fec03ac899cdc847f20e553de7850a83caaa71/search/binary_search.h) for example

**Comment by kvedala** [comment]
This looks much better, thank you.
Now, can you structure it in the doxygen-format as shown in [`search/binary_search.h`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/16fec03ac899cdc847f20e553de7850a83caaa71/search/binary_search.h) and described [here](https://www.doxygen.nl/manual/docblocks.html). This will help the system to automatically build documentation like this: https://kvedala.github.io/C-Plus-Plus without manual intervention.

**Comment by kvedala** [comment]
also, the comment blocks should go just above the function declaration

File: geometry/line_segment_intersection.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
+#include <iostream>
+
+struct Point {
+    int x, y;
+};
+
+struct SegmentIntersection {
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:7]: this struct is undocumented

File: geometry/line_segment_intersection.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
+#include <iostream>
+
+struct Point {
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:3]: struct is undocumented.
you can add a brief comment like so: [`others/smallest_circle`](https://github.com/kvedala/C-Plus-Plus/blob/master/others/smallest_circle.cpp)

File: geometry/line_segment_intersection.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    int x, y;
+};
+
+struct SegmentIntersection {
+    inline bool intersect(Point first_point, Point second_point,
+                                    Point third_point, Point forth_point) {
+       /**
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:10]: comment blocks should go above the function name i.e. for example, the comment block between lines 10-14 should be just above line 8.

File: geometry/line_segment_intersection.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                                                                forth_point))
+            return true;
+
+        else
+            return false;
+    }
+
+    inline int direction(Point first_point, Point second_point,
+                                                    Point third_point) {
+       /**
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:46]: same - comment block from lines 46-54 should be at line 44.

File: geometry/line_segment_intersection.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            std::min(first_point.y, second_point.y) <= third_point.y &&
+            third_point.y <= std::max(first_point.y, second_point.y))
+            return true;
+
+        else
+            return false;
+    }
+};
+
+int main() {
```
Comment (kvedala) [inline geometry/line_segment_intersection.cpp:76]: document the main function. a simple one-liner is good

**Comment by meSajied** [comment]
> also, the comment blocks should go just above the function declaration

this means after the include??

**Comment by meSajied** [comment]
Can you now merge this pull request??

**Comment by meSajied** [comment]
Can you merge this pull request now @kvedala @cclauss??

File: geometry/line_segment_intersection.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            third_point.x <= std::max(first_point.x, second_point.x) &&
+            std::min(first_point.y, second_point.y) <= third_point.y &&
+            third_point.y <= std::max(first_point.y, second_point.y))
+            return true;
+
+        else
+            return false;
+    }
+};
+/*
```
Comment (ayaankhan98) [inline geometry/line_segment_intersection.cpp:85]:
```suggestion
/**
```

File: geometry/line_segment_intersection.cpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
+/**
+ * @file
+ * @brief check whether two line segments intersect each other
+ * or not.
+ */
+#include <iostream>
+/*
```
Comment (ayaankhan98) [inline geometry/line_segment_intersection.cpp:8]:
```suggestion
/**
```

**Comment by meSajied** [comment]
Ithink everythink is good now.... @ayaankhan98

**Comment by kvedala** [comment]
Reason why we need a robust CI - [build fails on windows](https://github.com/kvedala/C-Plus-Plus/runs/724947972?check_suite_focus=true#step:5:232) due to missing `#include <algorothm>` for the functions `std::max` and `std::min`

**Comment by meSajied** [comment]
But now pull request is merged... @kvedala

**Comment by kvedala** [comment]
you can recreate a pull request with corrections. If you make a pull request - [here](https://github.com/kvedala/C-Plus-Plus), it will compile and check your code automatically

**Pull request merged**